File: julia/src/stabilizer_renyi_entropy.jl
"""
Stabilizer Rényi Entropy (SRE) for quantum states.

Implements the \u03b1-SRE

    M_\u03b1(|\u03c8\u27e9) = [log\u2082(\u03a3_P \u039e(P)^\u03b1) / (1 - \u03b1)] - log\u2082 d

where the characteristic function is

    \u039e(P) = |\u27e8\u03c8|P|\u03c8\u27e9|\u00b2 / 2^n.

For \u03b1 = 1 the Shannon-entropy limit is used:

    M_1 = -\u03a3_P \u039e(P) log\u2082 \u039e(P) - log\u2082 d,

with the convention 0 log 0 = 0.

A fast (unchecked) path is provided for hot loops over many states
with fixed Paulis and \u03b1; validate once outside the loop.
"""
module Magic

export stabilizer_renyi_entropy,
       stabilizer_renyi_entropy_unchecked,
       validate_alpha,
       validate_compatible,
       characteristic_function

using LinearAlgebra
using ..PauliAlgebra: PauliOperators
using ..QuantumStates: QuantumState, state_vector


# ============================================================================
# Validation helpers
# ============================================================================

"""
    validate_alpha(\u03b1::Real)

Validate the Rényi order \u03b1.

- \u03b1 \u2264 0        : throws `ArgumentError`.
- \u03b1 = 1        : issues a warning (Shannon-entropy limit).
- \u03b1 non-integer: issues a warning (generalised \u03b1-SRE).
- \u03b1 = 2, 3, \u2026  : valid, no warning.
"""
function validate_alpha(\u03b1::Real)
    \u03b1 <= 0 && throw(ArgumentError(
        "\u03b1 must be positive, got $(\u03b1). SRE is not defined for \u03b1 \u2264 0."
    ))
    if \u03b1 == 1
        @warn "\u03b1=1: computing M\u2081 directly (Shannon-entropy limit of the SRE)."
    elseif !isinteger(\u03b1)
        @warn "\u03b1=$(\u03b1) is not an integer. Computing the generalised \u03b1-Rényi entropy."
    end
end


"""
    validate_compatible(state::QuantumState, paulis::PauliOperators)

Throw an `ArgumentError` if `state.n_qubits \u2260 paulis.n_qubits`.
"""
function validate_compatible(state::QuantumState, paulis::PauliOperators)
    state.n_qubits != paulis.n_qubits && throw(ArgumentError(
        "state.n_qubits=$(state.n_qubits) \u2260 paulis.n_qubits=$(paulis.n_qubits)."
    ))
end


# ============================================================================
# Core computation
# ============================================================================

"""
    characteristic_function(\u03c8::AbstractVector{<:Complex}, paulis::PauliOperators) -> Vector{Float64}

Compute \u039e(P) = |\u27e8\u03c8|P|\u03c8\u27e9|\u00b2 / 2^n for every P in `paulis`.

Returns a `Vector{Float64}` of length `4^n` in label order.
The matrix-vector products `P * \u03c8` exploit sparsity via `SparseMatrixCSC`.
"""
function characteristic_function(
    \u03c8::AbstractVector{<:Complex},
    paulis::PauliOperators
)::Vector{Float64}
    norm_factor = 2^paulis.n_qubits
    return Float64[abs2(dot(\u03c8, P * \u03c8)) / norm_factor for (_, P) in paulis]
end


"""
    _compute_sre(\u03be::AbstractVector{Float64}, \u03b1::Real) -> Float64

Evaluate M_\u03b1 from the characteristic function vector \u03be.

For \u03b1 = 1 uses the Shannon-entropy formula with `0 log 0 := 0`.
"""
function _compute_sre(\u03be::AbstractVector{Float64}, \u03b1::Real)::Float64
    d = length(\u03be)
    if \u03b1 == 1
        return -sum(x > 0 ? x * log2(x) : 0.0 for x in \u03be) - log2(d)
    end
    return log2(sum(x^\u03b1 for x in \u03be)) / (1 - \u03b1) - log2(d)
end


# ============================================================================
# Unchecked fast path
# ============================================================================

"""
    stabilizer_renyi_entropy_unchecked(state, paulis, \u03b1 = 2; pauli_spectrum = false)

Fast-path SRE computation with no validation.

The caller is responsible for ensuring:
- `\u03b1 > 0`
- `state.n_qubits == paulis.n_qubits`
- `state` is a valid, normalised pure `QuantumState`

Intended for hot loops:

```julia
validate_alpha(\u03b1)
validate_compatible(states[1], paulis)
results = [stabilizer_renyi_entropy_unchecked(s, paulis, \u03b1) for s in states]
```

# Returns
- `Float64` if `pauli_spectrum = false` (default).
- `Tuple{Float64, Dict{String, Float64}}` if `pauli_spectrum = true`.
"""
function stabilizer_renyi_entropy_unchecked(
    state      ::QuantumState,
    paulis     ::PauliOperators,
    \u03b1          ::Real = 2;
    pauli_spectrum::Bool = false
)
    \u03be  = characteristic_function(state_vector(state), paulis)
    sre = _compute_sre(\u03be, \u03b1)
    pauli_spectrum || return sre
    spectrum = Dict(lbl => \u03be[i] for (i, (lbl, _)) in enumerate(paulis))
    return sre, spectrum
end


# ============================================================================
# Public validated entry point
# ============================================================================

"""
    stabilizer_renyi_entropy(state, paulis, \u03b1 = 2; pauli_spectrum = false)

Compute the \u03b1-Stabilizer Rényi Entropy for a pure quantum state.

Validated entry point. For hot loops over many states with fixed
`paulis` and `\u03b1`, validate once then call `stabilizer_renyi_entropy_unchecked`.

# Arguments
- `state::QuantumState`         \u2014 pure quantum state.
- `paulis::PauliOperators`      \u2014 Pauli operator set matching `state.n_qubits`.
- `\u03b1::Real`                    \u2014 Rényi order (default 2).
- `pauli_spectrum::Bool = false`\u2014 if true, also return the \u039e spectrum.

# Returns
- `Float64` if `pauli_spectrum = false`.
- `Tuple{Float64, Dict{String, Float64}}` if `pauli_spectrum = true`.

# Throws
- `ArgumentError` if `\u03b1 \u2264 0` or `state.n_qubits \u2260 paulis.n_qubits`.
"""
function stabilizer_renyi_entropy(
    state      ::QuantumState,
    paulis     ::PauliOperators,
    \u03b1          ::Real = 2;
    pauli_spectrum::Bool = false
)
    validate_alpha(\u03b1)
    validate_compatible(state, paulis)
    return stabilizer_renyi_entropy_unchecked(state, paulis, \u03b1; pauli_spectrum = pauli_spectrum)
end

end # module Magic

File: python/src/domcsis_epic_tinker_box/magic/stabilizer_renyi_entropy.py
"""
This module implements the Stabilizer Rényi Entropy (SRE) for quantum states.
"""


from __future__ import annotations

import warnings
import numpy as np
from typing import overload, Literal

from ..pauli_algebra.pauli_operators import PauliOperators
from ..quantum_state.quantum_state import QuantumState


# ============================================================================
# Overloads — public entry point
# ============================================================================

@overload
def stabilizer_renyi_entropy(
    state: QuantumState,
    paulis: PauliOperators,
    alpha: int | float = ...,
    pauli_spectrum: Literal[False] = ...,
) -> float: ...

@overload
def stabilizer_renyi_entropy(
    state: QuantumState,
    paulis: PauliOperators,
    alpha: int | float = ...,
    pauli_spectrum: Literal[True] = ...,
) -> tuple[float, dict[str, float]]: ...


# ============================================================================
# Public entry point — validated
# ============================================================================

def stabilizer_renyi_entropy(
    state: QuantumState,
    paulis: PauliOperators,
    alpha: int | float = 2,
    pauli_spectrum: bool = False,
) -> float | tuple[float, dict[str, float]]:
    """
    Compute the alpha-Stabilizer Rényi Entropy (SRE) for a pure quantum state.

    Validated entry point. For hot loops over many states with fixed paulis
    and alpha, validate once and use _stabilizer_renyi_entropy_unchecked
    directly to avoid redundant checks.

    The SRE is defined as:
        M_alpha(|psi>) = (1 / (1 - alpha)) * log( sum_{P in P_n} Xi(P)^alpha )
    where the characteristic function is:
        Xi(P) = |<psi|P|psi>|^2 / 2^n

    Special values of alpha:
        alpha=2 (default): Linear Magic M_2, the most common SRE variant.
        alpha=1:           Linear Magic M_1; a warning is issued.
        alpha>0, non-int:  general alpha-SRE; a warning is issued.
        alpha<=0:          undefined; raises ValueError.

    Args:
        state:
            Pure quantum state as a QuantumState instance. Normalisation,
            dimension, and purity are guaranteed by construction.
        paulis:
            PauliOperators instance. Must match state.n_qubits.
        alpha:
            Rényi order. Must be positive. Default is 2 (Linear Magic M_2).
        pauli_spectrum:
            If False (default), return only the SRE value.
            If True, return (SRE, spectrum) where spectrum maps each Pauli
            label to its characteristic function value Xi(P).

    Returns:
        float:
            The SRE value (if pauli_spectrum=False).
        tuple[float, dict[str, float]]:
            (SRE, Xi spectrum) if pauli_spectrum=True.

    Raises:
        ValueError: If state.n_qubits != paulis.n_qubits.
        ValueError: If alpha <= 0.
    """
    _validate_alpha(alpha)
    _validate_compatible(state, paulis)
    return _stabilizer_renyi_entropy_unchecked(state, paulis, alpha, pauli_spectrum)


# ============================================================================
# Fast path — unchecked
# ============================================================================

def _stabilizer_renyi_entropy_unchecked(
    state: QuantumState,
    paulis: PauliOperators,
    alpha: int | float = 2,
    pauli_spectrum: bool = False,
) -> float | tuple[float, dict[str, float]]:
    """
    Fast path SRE computation — no validation.

    Caller is responsible for ensuring:
        - alpha > 0
        - state.n_qubits == paulis.n_qubits
        - state is a valid QuantumState (normalised, pure)

    Intended for hot loops over many states with fixed paulis and alpha.
    Validate once with the public entry point, then call this directly:

        _validate_alpha(alpha)
        _validate_compatible(states[0], paulis)
        results = [
            _stabilizer_renyi_entropy_unchecked(state, paulis, alpha)
            for state in states
        ]

    Args:
        state:        QuantumState instance (unchecked).
        paulis:       PauliOperators instance (unchecked).
        alpha:        Rényi order (unchecked, assumed positive).
        pauli_spectrum: See stabilizer_renyi_entropy.

    Returns:
        float | tuple[float, dict[str, float]]: See stabilizer_renyi_entropy.
    """
    xi = _characteristic_function(state.vector, paulis)
    sre = _compute_sre(xi, alpha)
    if pauli_spectrum:
        return sre, dict(zip(paulis.labels, xi.tolist()))
    return sre


# ============================================================================
# Private helpers
# ============================================================================

def _validate_alpha(
    alpha: int | float,
) -> None:
    """
    Validate and characterise the Rényi order alpha.

    Rules:
        alpha <= 0:          raises ValueError (undefined).
        alpha == 1:          Linear Magic M_1; issues UserWarning.
        alpha == 2:          Linear Magic M_2; valid, no warning.
        alpha > 0, non-int:  valid; issues UserWarning.
        other positive int:  general alpha-SRE; valid, no warning.
    """
    if alpha <= 0:
        raise ValueError(
            f"alpha must be positive, got {alpha}. "
            "SRE is not defined for alpha <= 0."
        )
    if alpha == 1:
        warnings.warn(
            "alpha=1: computing Linear Magic M_1. "
            "Note that alpha -> 1 also recovers the Shannon entropy in the limit; "
            "here the finite formula is used directly.",
            UserWarning,
            stacklevel=3,
        )
    if not isinstance(alpha, int) and alpha != int(alpha):
        warnings.warn(
            f"alpha={alpha} is not an integer. "
            "Computing the general alpha-Rényi entropy. "
            "Ensure this is intentional.",
            UserWarning,
            stacklevel=3,
        )


def _validate_compatible(
    state: QuantumState, 
    paulis: PauliOperators,
) -> None:
    if state.n_qubits != paulis.n_qubits:
        raise ValueError(
            f"state.n_qubits={state.n_qubits} does not match "
            f"paulis.n_qubits={paulis.n_qubits}."
        )


def _characteristic_function(
    psi: np.ndarray,
    paulis: PauliOperators,
) -> np.ndarray:
    """
    Compute Xi(P) = |<psi|P|psi>|^2 / 2^n for all P in paulis.

    Args:
        psi:    Normalised state vector of shape (2^n,), from QuantumState.vector.
        paulis: PauliOperators instance for n qubits.

    Returns:
        np.ndarray of shape (4^n,) with Xi values in paulis label order.
    """
    norm = 2 ** paulis.n_qubits
    return np.array(
        [abs(psi.conj() @ (P @ psi)) ** 2 / norm for _, P in paulis.items()],
        dtype=float,
    )


def _compute_sre(
    xi: np.ndarray,
    alpha: int | float,
) -> float:
    """
    Compute M_alpha = log2( sum(Xi^alpha) ) / (1 - alpha) - log2(d).

    For alpha=1, the formula degenerates (division by zero) and the
    Shannon entropy limit is used instead:
        M_1 = -sum_P Xi(P) * log2(Xi(P)) - log2(d)
    where 0 * log2(0) := 0 by convention.

    Args:
        xi:    np.ndarray of characteristic function values Xi(P).
        alpha: Rényi order (positive).

    Returns:
        float: The SRE value.
    """
    d = xi.size
    if alpha == 1:
        # Shannon entropy limit: 0 * log2(0) = 0 by convention
        with np.errstate(divide="ignore"):
            log_xi = np.where(xi > 0, np.log2(xi), 0.0)
        return float(-np.dot(xi, log_xi) - np.log2(d))
    
    return float((np.log2(np.sum(xi ** alpha)) / (1 - alpha)) - np.log2(d))